File: neetcode150/python-solutions/2_pointers/three_sum.py
from typing import List


def three_sum(nums: List[int]) -> List[List[int]]:
    result: List[List[int]] = []
    nums.sort()

    for i, num in enumerate(nums):
        if num > 0:
            break
        if i > 0 and num == nums[i - 1]:
            continue

        l: int = i + 1
        r: int = len(nums) - 1
        while l < r:
            total = num + nums[l] + nums[r]
            if total < 0:
                l += 1
            elif total > 0:
                r -= 1
            else:
                result.append([num, nums[l], nums[r]])
                l += 1
                r -= 1
                while nums[l] == nums[l - 1] and l < r:
                    l += 1

    return result

File: neetcode150/rust-solutions/src/two_pointers/three_sum.rs
fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {

    let mut result: Vec<Vec<i32>> = Vec::new();

    nums.sort();

    for i in 0..nums.len() {
        if i > 0 && nums[i] == nums[i - 1] {
            continue;
        }

        let mut left = i + 1;
        let mut right = nums.len() - 1;

        while left < right {
            let total = nums[i] + nums[left] + nums[right];

            if total == 0 {
                result.push(vec![nums[i], nums[left], nums[right]]);
                left += 1;
                while nums[left] == nums[left - 1] && left < right {
                    left += 1;
                }
                right -= 1;
                while nums[right] == nums[right + 1] && left < right {
                    right -= 1;
                }
            } else if total < 0 {
               left += 1; 
            } else {
                right -= 1;
            }
        }
    }

    return result;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_three_sum_1() {
        let nums = vec![-1, 0, 1, 2, -1, 4];
        let got = three_sum(nums);
        let want = vec![vec![-1, -1, 2], vec![-1, 0, 1]];
        assert_eq!(got, want)
    }
}
